// @trpc/server/adapters/fastify/dist/trpc-server-adapters-fastify.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var url = require('url');
var resolveHTTPResponse = require('../../../dist/resolveHTTPResponse-3867844c.cjs.dev.js');
var adapters_ws_dist_trpcServerAdaptersWs = require('../../ws/dist/trpc-server-adapters-ws.cjs.dev.js');
require('../../../dist/transformTRPCResponse-8248515e.cjs.dev.js');
require('../../../dist/codes-33cef953.cjs.dev.js');
require('../../../dist/subscription-b8831081.cjs.dev.js');
require('events');

resolveHTTPResponse.assertNotBrowser();
async function fastifyRequestHandler(opts) {
  var _opts$req$body;

  const createContext = async function _createContext() {
    var _opts$createContext;

    return (_opts$createContext = opts.createContext) === null || _opts$createContext === void 0 ? void 0 : _opts$createContext.call(opts, opts);
  };

  const query = opts.req.query ? new url.URLSearchParams(opts.req.query) : new url.URLSearchParams(opts.req.url.split('?')[1]);
  const req = {
    query,
    method: opts.req.method,
    headers: opts.req.headers,
    body: (_opts$req$body = opts.req.body) !== null && _opts$req$body !== void 0 ? _opts$req$body : 'null'
  };
  const result = await resolveHTTPResponse.resolveHTTPResponse({
    req,
    createContext,
    path: opts.path,
    router: opts.router,
    batching: opts.batching,
    responseMeta: opts.responseMeta,

    onError(o) {
      var _opts$onError;

      opts === null || opts === void 0 ? void 0 : (_opts$onError = opts.onError) === null || _opts$onError === void 0 ? void 0 : _opts$onError.call(opts, { ...o,
        req: opts.req
      });
    }

  });
  const {
    res
  } = opts;

  if ('status' in result && (!res.statusCode || res.statusCode === 200)) {
    res.statusCode = result.status;
  }

  for (const [key, value] of Object.entries((_result$headers = result.headers) !== null && _result$headers !== void 0 ? _result$headers : {})) {
    var _result$headers;

    if (typeof value === 'undefined') {
      continue;
    }

    res.header(key, value);
  }

  res.send(result.body);
}

/// <reference types="fastify-websocket" />
function fastifyTRPCPlugin(fastify, opts, done) {
  var _opts$prefix;

  fastify.addContentTypeParser('application/json', {
    parseAs: 'string'
  }, function (_, body, _done) {
    _done(null, body);
  });
  fastify.all(`${(_opts$prefix = opts.prefix) !== null && _opts$prefix !== void 0 ? _opts$prefix : ''}/:path`, (req, res) => {
    const path = req.params.path;
    fastifyRequestHandler({ ...opts.trpcOptions,
      req,
      res,
      path
    });
  });

  if (opts.useWSS) {
    var _opts$prefix2;

    adapters_ws_dist_trpcServerAdaptersWs.applyWSSHandler({ ...opts.trpcOptions,
      wss: fastify.websocketServer
    }); // eslint-disable-next-line @typescript-eslint/no-empty-function

    fastify.get((_opts$prefix2 = opts.prefix) !== null && _opts$prefix2 !== void 0 ? _opts$prefix2 : '/', {
      websocket: true
    }, () => {});
  }

  done();
}

exports.fastifyRequestHandler = fastifyRequestHandler;
exports.fastifyTRPCPlugin = fastifyTRPCPlugin;
