// @trpc/server/adapters/next/dist/trpc-server-adapters-next.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var transformTRPCResponse = require('../../../dist/transformTRPCResponse-8248515e.cjs.dev.js');
var nodeHTTPRequestHandler = require('../../../dist/nodeHTTPRequestHandler-2cc444d8.cjs.dev.js');
require('url');
require('../../../dist/resolveHTTPResponse-3867844c.cjs.dev.js');
require('../../../dist/codes-33cef953.cjs.dev.js');

/* eslint-disable @typescript-eslint/no-explicit-any */
function createNextApiHandler(opts) {
  return async (req, res) => {
    function getPath() {
      if (typeof req.query.trpc === 'string') {
        return req.query.trpc;
      }

      if (Array.isArray(req.query.trpc)) {
        return req.query.trpc.join('/');
      }

      return null;
    }

    const path = getPath();

    if (path === null) {
      const error = opts.router.getErrorShape({
        error: new transformTRPCResponse.TRPCError({
          message: 'Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?',
          code: 'INTERNAL_SERVER_ERROR'
        }),
        type: 'unknown',
        ctx: undefined,
        path: undefined,
        input: undefined
      });
      const json = {
        id: -1,
        error
      };
      res.statusCode = 500;
      res.json(json);
      return;
    }

    await nodeHTTPRequestHandler.nodeHTTPRequestHandler({ ...opts,
      req,
      res,
      path
    });
  };
}

exports.createNextApiHandler = createNextApiHandler;
